File: vizdoom_experiment_statistics.r
# import packages
library(dplyr)
library(tidyr)
library(ggplot2)





# 1. I need to have the information like a list of the Subject numbers in the one file
# 2. Then it needs to do something for every number in that list 
    # 3. it should go into the corresponding folder
        # 4. it should do everything following for each blockfile
            # 5. only select the needed data
            # 6. make some descriptive Statistics (first action, correct action, optimal key, RT)
#7. have overall statistics

# 1. get the list of Subject numbers
sub_data <- read.csv("subjects.tsv", sep = "\t") #maybe work on avoiding duplicates in Subject.tsv

sub_list <- sub_data$sub_num
print(sub_list)
# do something for every number in that list

for (s in sub_list){
  # 3. navigate to the corresponding folder
  current_filepath <- file.path(getwd(), paste0("sub_",s))

  if (file.exists(current_filepath)) {
    #print(current_filepath)
    setwd(current_filepath) #setting the current working directory to access the files
    print(paste("Current working directory for sub_num", s, ":", getwd()))
    # do everything for each blockfile
    block_files <- list.files(pattern = "\\.tsv")

    for (file in block_files){
      current_blockfile <- read.csv(file, sep = "\t") #read the tsv-file
      needed_data <- current_blockfile %>%
        select(Subject, Block, Episode,State,Tic, Time, Action, Movement, Variation, X1_y, X1_name)  # Only keep relevant columns
      # ensure specific data is the right datatype  
      needed_data$Subject <- as.integer(needed_data$Subject) #sub number
      needed_data$Block <- as.integer(needed_data$Block) #block number
      needed_data$Episode <- as.integer(needed_data$Episode) #episode as int
      needed_data$State <- as.integer(needed_data$State) #state as int
      needed_data$Variation <- as.integer(needed_data$Variation) #variation as int
      needed_data$Movement <- as.integer(needed_data$Movement) #movement as int
      needed_data$Tic <- as.integer(needed_data$Tic)# tic as int
      needed_data$Time <- as.numeric(needed_data$Time)# time as numeric
      needed_data$X1_y <- as.numeric(needed_data$X1_y)# X1_y as numeric
      needed_data <- needed_data %>%
        filter(Subject != "invalid number" | is.na(X1_y))
      #View(needed_data)

      #all the functions needed for further processing
      # function to translate if movement was inverted or normal

      movement_translation <- function(movement){
        if (movement == 0) {
          return("normal")
        } else if (movement == 1) {
          return("inverted")
        } else {
          return("unknown")  # You can handle other cases if needed
        }
      }

      target_side <- function(position){
        if (is.na(position)){
          return ("unkown")
        }
        #if (is.numeric(position)) {
        position <- round(as.numeric(position), 2)
        #}  # Round to 2 decimal places
        if (position >= (75)) {
          return("left")
        }else if (position < (-9)){
          return("right")
        }else{
          return("error")
        }
      }

      #function to check if the movement choice was correct

      correct_choice <- function(movement,position, action){
        if (position == "left" && action == "MOVE_LEFT"){ #left side, MOVE_LEFT is correct
          return(TRUE)
        }else if (position == "right" && action == "MOVE_RIGHT"){ # right side, MOVE_RIGHT is correct
          return(TRUE)
        }else {   # everything else is incorrect
          return(FALSE)
        }

      }

      #function to check, if the movement was the optimal choice (choice with the overall highest probability)
      # if variation is 1 or 2 either Key A or D is the optimal choice, which  
      # So e.g. TRUE if variation == 1 and action = MOVE_LEFT optimal

      optimal_choice <- function(variation,movement,first_action){
        if (variation == 1 && movement == 0 && first_action == "MOVE_LEFT"){  # variation 1 (A optimal), movement normal and first action is move left
          return(TRUE)
        }else if (variation == 1 && movement == 1 && first_action == "MOVE_RIGHT"){ # variation 1 (A optimal), movement inverted and first action move right 
          return(TRUE)
        }else if (variation == 2 && movement == 0 && first_action == "MOVE_RIGHT"){ # variation 2 (D optimal), movement normal and first action move right
          return(TRUE)
        }else if (variation == 2 && movement == 1 && first_action == "MOVE_LEFT"){ # variation 2 (D optimal), movement inverted and first action move left
          return(TRUE)
        }else {
          return(FALSE)
        }
      }

      # Group the blockfiles by episode
      processed_data <- needed_data %>%
        group_by(Episode) %>%
    
        # Filter to get the first non-NA Action in each group
        summarize(
          First_Action = first(na.omit(Action)),  # Due to this, all episodes where the subjects do not move will be filtered!
      
          # Get the index of the first non-NA Action
          first_non_na_index = which(!is.na(Action))[1],
      
          Movement = if (!is.na(first_non_na_index) && first_non_na_index > 0) {
            Movement[first_non_na_index]
          } else {
            NA
          },
      
          Movement_Words = if (!is.na(Movement) && !is.na(first_non_na_index)) {
            movement_translation(Movement)
          } else {
            NA
          },
      
          Variation = if (!is.na(first_non_na_index) && first_non_na_index > 0) {
            Variation[first_non_na_index]
          } else {
            NA
          },
      
          X1_y = if (!is.na(first_non_na_index) && first_non_na_index > 0) {
            first(na.omit(X1_y))#X1_y[first_non_na_index] # need to change. in case of movement before target appears this value is NA either way
          } else {
            NA
          },
      
          side = if (!is.na(first_non_na_index) && first_non_na_index > 0) {
            target_side(X1_y)
          } else {
            NA
          },

          correct_choice = if (!is.na(first_non_na_index) && first_non_na_index > 0) {
            correct_choice(side, First_Action)
          } else {
            NA
          },

          optimal_choice = if (!is.na(first_non_na_index) && first_non_na_index > 0) {
            optimal_choice(Variation, Movement, First_Action)
          } else {
            NA
          },
      
          Tic = if (!is.na(first_non_na_index) && first_non_na_index > 0) {
            Tic[first_non_na_index]
          } else {
            NA
          },
      
          Time_of_Action = if (!is.na(first_non_na_index) && first_non_na_index > 0) {
            Time[first_non_na_index]
          } else {
            NA
          },
          Time_of_Episode = max(Time),

          #Was the traget hit?
          Hit = any(grepl("^Dead", X1_name))
        ) %>%
        filter(if_all(everything(), ~ . != "invalid number")) %>% #kick out all the rows with invalid numbers
        ungroup()
      # Reorder the columns so Subject and Block come before Episode
      processed_data <- processed_data %>%
        mutate(Subject = s, Block = as.numeric(gsub("\\D", "", file))) %>%
        select(Subject, Block, everything())
      # View the processed data
      # create folders for processed data if they don't exist and setwd()
      processed_folder <- "processed data"
      if (!file.exists(processed_folder)){
        dir.create(file.path(getwd(),processed_folder))
      }
      setwd(file.path(getwd(),processed_folder))
      
      
      write.table(processed_data, paste0("processed_data_",file), row.names = FALSE, sep = "\t")
      
      setwd("..")
    }

    #at the very end set the working directory back to the original
    setwd("..")
    print(paste("Returned to original working directory:", getwd()))
  
  } else {
    print(paste("Skipping sub_num", s, "because folder does not exist"))
  }
}
# now I have files that are processed and so far contain the data I need for 
# my descriptive statistics
# what I want to have is:
# RT
# optimal choices and relations
# correct choices
# episode length:Time_of_Episode mean (per block, per sub, overall)
# basically I have to store all the data of the files in one file at least for the numbers I want overall 
# 1. for all the Subjects go to the folder of the processed data
# 2. for each file
# 3. read the file
# 4. append it to a dataframe
# 5. save that dataframe 


# create dataframe for all data

descriptive_statistics_overall <- data.frame()
for (s in sub_list){
  folderpath <- file.path(getwd(),paste0("sub_",s),"processed data")
  if (file.exists(folderpath)) {
    setwd(folderpath)
    print(paste("current path: ", getwd()))
    s_files <- list.files(full.names = TRUE)
    print(s_files)

    for (f in s_files){
      current_f <- read.table(f, header = TRUE)
      #print(paste("Processing file: ", f))  
      descriptive_statistics_overall <- rbind(descriptive_statistics_overall,current_f)
    }

    #at the end set back to basic directory
    setwd("../..")
  } else {
    print(paste("Skipping processed data for sub_num", s, "because folder does not exist"))
  }
}

descriptive_statistics_overall <- descriptive_statistics_overall %>%
  group_by(Subject, Block) %>%
    mutate(cum_opt = cumsum(as.numeric(optimal_choice)), .after = optimal_choice)
View(descriptive_statistics_overall)

#get the overall statistics

overall_statistics <- descriptive_statistics_overall %>%
  ungroup() %>%
  summarize(
    mean_time_episode = mean(Time_of_Episode),
    sd_time_episode = sd(Time_of_Episode),
    mean_RT_first_action = mean(Time_of_Action),
    sd_RT_first_action = sd(Time_of_Action),
    optimal_choice_count = (sum(optimal_choice == TRUE)/n()),
    correct_choice_count = (sum(correct_choice == TRUE)/n())
  )

View(overall_statistics)

# Get the overall ratio statistics

ratio_statistics <- descriptive_statistics_overall %>%
  group_by(Variation) %>%
    summarize(
      normal_right = sum(side == "right" & Movement == 0)/n(),
      inverted_right = sum(side == "right" & Movement == 1)/n(), 
      normal_left = sum(side == "left" & Movement == 0)/n(), 
      inverted_left = sum(side == "left" & Movement == 1)/n()

    )
View(ratio_statistics)


# Get Block statistics for each Subject
block_statistics <- descriptive_statistics_overall %>%
  group_by(Subject, Block, Variation) %>%
  summarize(
    episodes = n(),
    mean_time_episode = mean(Time_of_Episode),
    sd_time_episode = sd(Time_of_Episode),
    mean_RT_first_action = mean(Time_of_Action),
    sd_RT_first_action = sd(Time_of_Action),
    optimal_choice_count = (sum(optimal_choice == TRUE)/n()),
    correct_choice_count = (sum(correct_choice == "correct")/n())
  )

View(block_statistics)

# If the variation is 1 there's a probability of .8, the movement is
# inverted if the target spawns on the right side (optimal choice A/left)
# and a probability of .8 that the movement is normal 
# if the target spawns on the left side (optimal choice A/left)
# If variation 2 prob of .8 that normal if right and prob .8
# that inv if left 

#to check the ratios and also if on a total it was close to
#the goal (.8/.2) the ratios of the .8 should be checked so 
#4 conditions (2 per balance value) should be summed and ratioed

# Get block-wise ratio statistics for each Subject
ratio_blockwise <- descriptive_statistics_overall %>%
  group_by(Subject, Block) %>%
  summarize(
    Variation = first(Variation),  # Assuming Balance is the same for each Subject-Block combination
    normal_right = sum(side == "right" & Movement == 0) / n(),
    inverted_right = sum(side == "right" & Movement == 1) / n(),
    normal_left = sum(side == "left" & Movement == 0) / n(),
    inverted_left = sum(side == "left" & Movement == 1) / n(),
    times_left = sum(side=="left"),
    times_right = sum(side=="right"),
    Episodes = n()
    #.groups = 'drop'  # Optional: This prevents grouping in the output
  )

View(ratio_blockwise)




##### Plots

#plot Episode time (do they get faster within and over blocks?)
#to consider: the target spawns in random places therefore the distance changes
#which then changes the time necessary to finish the episode
#maybe there's a way to normalize it (set it in relation to distance)?

# mean Episode times for each Subject and block
time_data <- descriptive_statistics_overall %>%
  group_by(Subject, Block) %>%
  summarise(
    mean_ep_time_sub = mean(Time_of_Episode, na.rm = TRUE), # mean ep-time per block for each Subject
    .groups = "drop"
  )

#  mean Episode times for each block
mean_times_all <- time_data %>%
  group_by(Block) %>%
  summarise(
    mean_ep_time_all = mean(mean_ep_time_sub, na.rm = TRUE),
    .groups = "drop"
  )

# plot for Episode times per block, mean and individual
episode_time_plot <- ggplot() +
  # Individual proportions
  geom_point(data = time_data, 
             aes(x = Block, y = mean_ep_time_sub, colour = factor(Subject)), 
             alpha = 0.6, size = 2) + 
  # Mean proportions
  geom_line(data = mean_times_all, 
            aes(x = Block, y = mean_ep_time_all, group = 1), 
            colour = "black", size = 1) +
  geom_point(data = mean_times_all, 
             aes(x = Block, y = mean_ep_time_all), 
             colour = "black", size = 3) +
  labs(
    x = "Block", 
    y = "Episode Time", 
    colour = "Subject",
    title = "Individual and overall mean times of episodes per Block"
  ) +
  theme_minimal() +
  # Customize text sizes using `theme()`
  theme_minimal(base_size = 12) +  # Adjust overall base text size
  theme(
    plot.title = element_text(size = 18, face = "bold"),  # Title size and style
    axis.title = element_text(size = 14),  # Axis labels size
    axis.text = element_text(size = 12),  # Axis tick labels size
    legend.text = element_text(size = 12),  # Legend text size
    legend.title = element_text(size = 14)  # Legend title size
  )

#Save it
ggsave("Episode_Times.pdf", plot = episode_time_plot,
        width = 210, height = 210, units = "mm")


# plot for the RT (time of first movement)

# mean reaction times for each Subject and block
RT_data <- descriptive_statistics_overall %>%
  group_by(Subject, Block) %>%
  summarise(
    mean_RT_sub = mean(Time_of_Action, na.rm = TRUE), # mean RT per block for each Subject
    .groups = "drop"
  )

# Calculate mean Episode times for each block
mean_RT_all <- RT_data %>%
  group_by(Block) %>%
  summarise(
    mean_RT_all = mean(mean_RT_sub, na.rm = TRUE),
    .groups = "drop"
  )

# RT-plot

RT_plot <- ggplot() +
  # Individual means
  geom_point(data = RT_data, 
             aes(x = Block, y = mean_RT_sub, colour = factor(Subject)), 
             alpha = 0.6, size = 2) + 
  # Means for all
  geom_line(data = mean_RT_all, 
            aes(x = Block, y = mean_RT_all, group = 1), 
            colour = "black", size = 1) +
  geom_point(data = mean_RT_all, 
             aes(x = Block, y = mean_RT_all), 
             colour = "black", size = 3) +
  labs(
    x = "Block", 
    y = "Reaction Time", 
    colour = "Subject",
    title = "Individual and overall Reaction Times per Block"
  ) +
  theme_minimal() +
  # Customize text sizes using `theme()`
  theme_minimal(base_size = 12) +  # Adjust overall base text size
  theme(
    plot.title = element_text(size = 18, face = "bold"),  # Title size and style
    axis.title = element_text(size = 14),  # Axis labels size
    axis.text = element_text(size = 12),  # Axis tick labels size
    legend.text = element_text(size = 12),  # Legend text size
    legend.title = element_text(size = 14)  # Legend title size
  )


#Save it
ggsave("RT_A4.pdf", plot = RT_plot,
        width = 210, height = 210, units = "mm")




# Calculate proportions for each Subject and block
proportions_data <- descriptive_statistics_overall %>%
  group_by(Subject, Block) %>%
  summarise(
    prop_opt = mean(optimal_choice, na.rm = TRUE), # Proportion for each Subject
    .groups = "drop"
  )

# Calculate mean proportions for each block
mean_proportions <- proportions_data %>%
  group_by(Block) %>%
  summarise(
    mean_prop_opt = mean(prop_opt, na.rm = TRUE),
    .groups = "drop"
  )

# plot for optimal choice proportion per block, mean and individual
proportion_plot <- ggplot() +
  # Individual proportions
  geom_point(data = proportions_data, 
             aes(x = Block, y = prop_opt, colour = factor(Subject)), 
             alpha = 0.6, size = 2) + # , position = position_jitter(width = 0.2) Jitter to avoid overlap
  # Mean proportions
  geom_line(data = mean_proportions, 
            aes(x = Block, y = mean_prop_opt, group = 1), 
            colour = "black", size = 1) +
  geom_point(data = mean_proportions, 
             aes(x = Block, y = mean_prop_opt), 
             colour = "black", size = 3) +
  labs(
    x = "Block", 
    y = "Proportion of Optimal Choices", 
    colour = "Subject",
    title = "Proportion of Optimal Choices by Block"
  ) +
  theme_minimal()+
  # Customize text sizes using `theme()`
  theme_minimal(base_size = 12) +  # Adjust overall base text size
  theme(
    plot.title = element_text(size = 18, face = "bold"),  # Title size and style
    axis.title = element_text(size = 14),  # Axis labels size
    axis.text = element_text(size = 12),  # Axis tick labels size
    legend.text = element_text(size = 12),  # Legend text size
    legend.title = element_text(size = 14)  # Legend title size
  )
# Save

# Save the plot
ggsave("Proportion_Optimal_Choices.pdf", 
       plot = proportion_plot, 
       width = 210, height = 210, units = "mm")



# Calculate proportions for each Subject and block (correct choice)
proportions_data2 <- descriptive_statistics_overall %>%
  group_by(Subject, Block) %>%
  summarise(
    prop_cor = mean(correct_choice, na.rm = TRUE), # Proportion for each Subject
    .groups = "drop"
  )

# Calculate mean proportions for each block
mean_proportions_cor <- proportions_data2 %>%
  group_by(Block) %>%
  summarise(
    mean_prop_cor = mean(prop_cor, na.rm = TRUE),
    .groups = "drop"
  )

# plot for correct choice proportion per block, mean and individual
proportion_plot_cor <- ggplot() +
  # Individual proportions
  geom_point(data = proportions_data2, 
             aes(x = Block, y = prop_cor, colour = factor(Subject)), 
             alpha = 0.6, size = 2) + # , position = position_jitter(width = 0.2) Jitter to avoid overlap
  # Mean proportions
  geom_line(data = mean_proportions_cor, 
            aes(x = Block, y = mean_prop_cor, group = 1), 
            colour = "black", size = 1) +
  geom_point(data = mean_proportions_cor, 
             aes(x = Block, y = mean_prop_cor), 
             colour = "black", size = 3) +
   # Labels and title
  labs(
    x = "Block", 
    y = "Proportion of Correct Choices", 
    colour = "Subject",
    title = "Proportion of Correct Choices by Block"
  ) +
  # Customize text sizes using `theme()`
  theme_minimal(base_size = 12) +  # Adjust overall base text size
  theme(
    plot.title = element_text(size = 18, face = "bold"),  # Title size and style
    axis.title = element_text(size = 14),  # Axis labels size
    axis.text = element_text(size = 12),  # Axis tick labels size
    legend.text = element_text(size = 12),  # Legend text size
    legend.title = element_text(size = 14)  # Legend title size
  )
# Save the plot
ggsave("Proportion_Correct_Choices.pdf", 
       plot = proportion_plot_cor, 
       width = 210, height = 210, units = "mm")

File: vizdoom_experiment_statistics.py
import pandas as pd
import os
from io import StringIO
import numpy as np

# Column names and data type conversions as before
columns_needed = [
    "Subject", "Block", "Episode", "State", "Tic", "Time", "Action", 
    "Movement", "Variation", "Target_pos", "Target_name", "Reward"
]

convert_to_int = ["Subject", "Block", "Episode", "State", "Tic", "Movement", "Variation"]
convert_to_float = ["Time", "Target_pos", "Reward"]

# Subject data
subject_file = "subjects.tsv"
sub_data = pd.read_csv(subject_file, sep="\t")
sub_list = sub_data["sub_num"].unique()

root_dir = os.getcwd()

def movement_translation(movement):
    if movement == 0:
        return "normal"
    elif movement == 1:
        return "inverted"
    else:
        return "unknown"  # For other cases

def target_side(position):
    if position is None:
        return "unknown"
    
    position = round(float(position), 2)  # Round to 2 decimal places
    if position >= 75:
        return "left"
    elif position < -9:
        return "right"
    else:
        return "error"

def correct_choice(position, action):
    if position == "left" and action == "MOVE_LEFT":
        return True
    elif position == "right" and action == "MOVE_RIGHT":
        return True
    else:
        return False

def optimal_choice(variation, movement, first_action):
    if variation == 1 and movement == 0 and first_action == "MOVE_LEFT":
        return True
    elif variation == 1 and movement == 1 and first_action == "MOVE_RIGHT":
        return True
    elif variation == 2 and movement == 0 and first_action == "MOVE_RIGHT":
        return True
    elif variation == 2 and movement == 1 and first_action == "MOVE_LEFT":
        return True
    else:
        return False

def hit_logic(df):
    # Check if Reward is 1 at any point in the episode as indicator that target has been hit
    if "Reward" in df.columns:
        return df["Reward"].eq(1.0).any()
    return False

def process_data(needed_data):
    def first_non_na(series):
        return series.dropna().iloc[0] if not series.dropna().empty else np.nan

    # how many episodes where recorded and replayed
    original_counts = (
        needed_data.groupby(["Subject", "Block", "Episode"]).size()
        .reset_index()
        .groupby(["Subject", "Block"])
        .size()
        .reset_index(name="Original_Episode_Count")
    )

    # only process valid episodes (excludes episodes where neither action nor target_pos have at least 1 value)
    valid_episodes = needed_data.groupby("Episode").filter(
        lambda g: g["Action"].notna().any() and g["Target_pos"].notna().any()
    )

    # processing of valid episodes
    processed_data = (
        valid_episodes.groupby("Episode").apply(lambda group: pd.Series({
            "Subject": group["Subject"].iloc[0],
            "Block": group["Block"].iloc[0],
            "Episode": group["Episode"].iloc[0],
            "State": group["State"].iloc[0],
            "Tic": group["Tic"].iloc[0],
            "Time": group["Time"].iloc[0],
            "Action": group["Action"].iloc[0],
            "Movement": group["Movement"].iloc[0],
            "Movement_Words": movement_translation(group["Movement"].iloc[0]),
            "Variation": group["Variation"].iloc[0],
            "Target_pos": first_non_na(group["Target_pos"]),
            "Target_name": first_non_na(group["Target_name"]),
            "first_non_na_index": group["Action"].first_valid_index(),
            "First_Action": first_non_na(group["Action"]),
            "Side": target_side(first_non_na(group["Target_pos"])),
            "correct_choice": correct_choice(target_side(first_non_na(group["Target_pos"])), first_non_na(group["Action"])),
            "optimal_choice": optimal_choice(
                group["Variation"].iloc[0],
                group["Movement"].iloc[0],
                first_non_na(group["Action"])
            ),
            "Time_of_Action": group.loc[group["Action"].first_valid_index(), "Time"],
            "Time_of_Episode": group["Time"].max(),
            "Hit": hit_logic(group)
        }))
        .reset_index(drop=True)
    )

    # counts valid episodes
    valid_counts = processed_data.groupby(["Subject", "Block"]).size().reset_index(name="Valid_Episode_Count")

    # merging
    overview = pd.merge(original_counts, valid_counts, on=["Subject", "Block"], how="left").fillna(0)
    overview["Valid_Episode_Count"] = overview["Valid_Episode_Count"].astype(int)
    overview["Excluded_Episodes"] = overview["Original_Episode_Count"] - overview["Valid_Episode_Count"]

    return processed_data, overview


all_subjects_combined = []
all_overviews = []  # episodecount

# Process each subject
for subject in sub_list:
    subject_folder = os.path.join(root_dir, f"sub_{subject}")
    if not os.path.exists(subject_folder):
        continue
    
    processed_folder = os.path.join(subject_folder, "processed_data")
    os.makedirs(processed_folder, exist_ok=True)

    # Create a container to hold processed blocks for each subject
    all_processed_data = []
     
    
    for file in os.listdir(subject_folder):
        if file.endswith(".tsv"):
            file_path = os.path.join(subject_folder, file)
            
            df_list = []
            with open(file_path, "r") as f:
                header = None
                for line in f:
                    if header is None or not line.startswith(header[0]):
                        df_list.append(line)
                    if header is None:
                        header = line.strip().split("\t")
            
            df = pd.read_csv(StringIO("".join(df_list)), sep="\t", engine="python", on_bad_lines="skip")
            
            # Process and convert data types as before
            df = df[[col for col in columns_needed if col in df.columns]]
            for col in convert_to_int:
                if col in df.columns:
                    df[col] = pd.to_numeric(df[col], errors='ignore')
            for col in convert_to_float:
                if col in df.columns:
                    df[col] = pd.to_numeric(df[col], errors='ignore')

            # Process each block and store in the all_processed_data container
            for block, block_df in df.groupby("Block"):
                processed_block_data, overview_df = process_data(block_df)  
                all_processed_data.append(processed_block_data)
                all_overviews.append(overview_df)
            
    # After processing all blocks for the subject, combine the results
    combined_data = pd.concat(all_processed_data, ignore_index=True)
    combined_data = combined_data.sort_values(by=["Block", "Episode"]).reset_index(drop=True)


    # Save the combined data to a single output file for this subject
    output_file = os.path.join(processed_folder, f"processed_data_subject_{subject}.tsv")
    combined_data.to_csv(output_file, sep="\t", index=False)
    
    print(f"Processed data saved for Subject {subject}: {output_file}")

    all_subjects_combined.append(combined_data)

# Combine all subjects' data into one file
if all_subjects_combined:
    grand_combined_data = pd.concat(all_subjects_combined, ignore_index=True)
    grand_combined_data = grand_combined_data.sort_values(by=["Subject", "Block", "Episode"]).reset_index(drop=True)
    
    final_output_file = os.path.join(root_dir, "all_subjects_combined.tsv")
    grand_combined_data.to_csv(final_output_file, sep="\t", index=False)
    print(f"All subject data combined into one file: {final_output_file}")

# Combine overview over all Subjects and Blocks
if all_overviews:
    combined_overview = pd.concat(all_overviews, ignore_index=True)
    combined_overview = combined_overview.sort_values(by=["Subject", "Block"]).reset_index(drop=True)

    print("\n📊 Episodecount pro Subject & Block:")
    print(combined_overview)

# Calculate the movement ratios for each variation to check if there was an overall ratio of .8/.2 
ratio_statistics = grand_combined_data.groupby("Variation").apply(lambda df: pd.Series({
    "normal_right": ((df["Side"] == "right") & (df["Movement"] == 0)).sum() / len(df),
    "inverted_right": ((df["Side"] == "right") & (df["Movement"] == 1)).sum() / len(df),
    "normal_left": ((df["Side"] == "left") & (df["Movement"] == 0)).sum() / len(df),
    "inverted_left": ((df["Side"] == "left") & (df["Movement"] == 1)).sum() / len(df),
})).reset_index()

print("\n✅ Ratio statistics:")
print(ratio_statistics)

# save ratio-statistics and episode-overview in one file

#  axis=1 for horizontal merge
combined_data = pd.concat([combined_overview, ratio_statistics], axis=1)

combined_data.to_csv("combined_statistics_overview.tsv", sep="\t", index=False)



import matplotlib.pyplot as plt

# Add a trial index per block (0-based within each Subject/Block)
grand_combined_data["Trial_Index"] = (
    grand_combined_data
    .groupby(["Subject", "Block"])
    .cumcount()
)

# Group by trial index and get mean optimal choice rate
trial_avg_optimal = (
    grand_combined_data
    .groupby("Trial_Index")["optimal_choice"]
    .mean()
)

# Add a Trial counter per block so we can align episodes across blocks
grand_combined_data["Trial"] = grand_combined_data["Episode"]


# Now compute the average proportion of optimal choices per trial
optimal_per_trial = (
    grand_combined_data
    .groupby("Trial")
    .agg(Optimal_Proportion=("optimal_choice", "mean"))
    .reset_index()
)




# plotting
plt.figure(figsize=(12, 6))
plt.plot(optimal_per_trial["Trial"], optimal_per_trial["Optimal_Proportion"], marker='o')
plt.xlabel("Trial")
plt.ylabel("Proportion of Optimal Choices")
plt.title("Average Optimal Choices per Trial")

# Fix x-axis ticks to show only integers
plt.xticks(ticks=range(1, optimal_per_trial["Trial"].max() + 1))

plt.grid(True)
plt.tight_layout()
plt.savefig("Average_optimal_choices_by_trial.pdf", dpi=300)
plt.close()


import matplotlib.pyplot as plt
import numpy as np

# Compute subject-level proportions
subject_trial_optimal = (
    grand_combined_data
    .groupby(["Subject", "Trial"])
    .agg(
        Optimal_Count=("optimal_choice", "sum"),
        Trial_Count=("optimal_choice", "count")
    )
    .reset_index()
)

subject_trial_optimal["Optimal_Proportion"] = (
    subject_trial_optimal["Optimal_Count"] / subject_trial_optimal["Trial_Count"]
)

# Compute trial-level means (already in your code)
optimal_per_trial = (
    grand_combined_data
    .groupby("Trial")
    .agg(Optimal_Proportion=("optimal_choice", "mean"))
    .reset_index()
)


# Set up plot
plt.figure(figsize=(12, 6))

# Plot subject dots with manual jitter
np.random.seed(0)  # for reproducibility
jitter_strength = 0.15

for _, row in subject_trial_optimal.iterrows():
    x = row["Trial"] + np.random.uniform(-jitter_strength, jitter_strength)
    y = row["Optimal_Proportion"]
    plt.plot(x, y, 'o', markerfacecolor=None,markeredgecolor='black', alpha=0.5, markersize=3)

# Plot average line
plt.plot(
    optimal_per_trial["Trial"],
    optimal_per_trial["Optimal_Proportion"],
    marker='o',
    color="blue",
    label="Mean Optimal Choice",
    zorder=2
)

# Final touches
plt.xlabel("Trial")
plt.ylabel("Proportion of Optimal Choices")
plt.title("Average Optimal Choices per Trial (with Subject Data)")
plt.xticks(ticks=range(1, optimal_per_trial["Trial"].max() + 1))
plt.grid(True)
plt.tight_layout()
plt.legend()
plt.savefig("Average_optimal_choices_by_trial_with_subjects_matplotlib.pdf", dpi=300)
plt.close()



# Count datapoints per trial
trial_counts = grand_combined_data.groupby("Trial").size().reset_index(name="Count")

# Plot with matplotlib
plt.figure(figsize=(12, 6))
plt.bar(trial_counts["Trial"], trial_counts["Count"], color="steelblue")
plt.title("Number of Data Points per Trial (Episode)")
plt.xlabel("Trial (Episode Number)")
plt.ylabel("Number of Data Points")
plt.xticks(trial_counts["Trial"], rotation=90)
plt.tight_layout()
plt.savefig("datapoints_per_trial.pdf", dpi=300)
plt.close()





# Count total and optimal per movement type
movement_group = grand_combined_data.groupby("Movement_Words")["optimal_choice"].agg(
    total='count', optimal='sum'
).reset_index()

# Calculate proportion
movement_group["optimal_rate"] = movement_group["optimal"] / movement_group["total"]

# Plot
plt.figure(figsize=(6, 5))
plt.bar(movement_group["Movement_Words"], movement_group["optimal_rate"], color=["skyblue", "salmon"])
plt.xlabel("Movement Type")
plt.ylabel("Proportion of Optimal Choices")
plt.title("Optimal Choice Rate by Movement Type")
plt.ylim(0, 1)
plt.grid(axis='y', linestyle='--', alpha=0.7)
plt.tight_layout()
plt.savefig("Optimal_choice_by_movement_type.pdf", dpi=300)
plt.close()

# Count total and optimal per block
block_group = grand_combined_data.groupby("Block")["optimal_choice"].agg(
    total='count', optimal='sum'
).reset_index()

# Calculate proportion
block_group["optimal_rate"] = block_group["optimal"] / block_group["total"]

# Plot
plt.figure(figsize=(8, 5))
plt.bar(block_group["Block"], block_group["optimal_rate"], color="mediumseagreen")
plt.xlabel("Block")
plt.ylabel("Proportion of Optimal Choices")
plt.title("Optimal Choice Rate by Block")
plt.ylim(0, 1)
plt.grid(axis='y', linestyle='--', alpha=0.7)
plt.tight_layout()
plt.savefig("Optimal_choice_by_block.pdf", dpi=300)
plt.close()

